File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys

N, M, V = map(int, sys.stdin.readline().rstrip().split())
graph = [[0]*(N+1) for _ in range(N+1)]
for _ in range(M):
    x, y = map(int, sys.stdin.readline().rstrip().split())
    graph[x][y] = 1
    graph[y][x] = 1
stack = [V]
d_v = [0]*(N+1)
queue = [V]
b_v = [0]*(N+1)
def dfs():
    while stack:
        k = stack.pop()
        if d_v[k] == 0:
            print(k, end=' ')
        d_v[k] = 1
        for l in range(N, -1, -1):
            if d_v[l] == 0 and graph[k][l] == 1:
                stack.append(l)

def bfs():
    while queue:
        k = queue.pop(0)
        if b_v[k] == 0:
            print(k, end=' ')
        b_v[k] = 1
        for l in range(N + 1):
            if b_v[l] == 0 and graph[k][l] == 1:
                queue.append(l)

dfs()
print()
bfs()

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {

    
    private static StringBuilder sb = new StringBuilder();
    private static int n, m, v;
    private static int[][] graph;
    private static boolean[] visited;
    private static ArrayDeque<Integer> queue = new ArrayDeque<Integer>();

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st1 = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st1.nextToken());
        m = Integer.parseInt(st1.nextToken());
        v = Integer.parseInt(st1.nextToken());


        graph = new int[n+1][n+1];
        for (int i = 0; i < m; i++) {
            StringTokenizer st2 = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st2.nextToken());
            int b = Integer.parseInt(st2.nextToken());
            graph[a][b] = graph[b][a] = 1;
        }
        visited = new boolean[n+1];
        dfs(v);
        sb.append("\n");

        visited = new boolean[n+1];
        bfs(v);

        bw.write(sb.toString());
        bw.flush();

        br.close();
        bw.close();
    }

    private static void dfs(int v) {
        visited[v] = true;
        sb.append(v+" ");
        for(int i=1; i<n+1; i++) {
            if (graph[v][i] == 1 && !visited[i]) {
                dfs(i);
            }
        }    
    }

    private static void bfs(int v) {
        queue.add(v);
        visited[v] = true;
        
        while(!queue.isEmpty()) {
            int x = queue.poll();
            sb.append(x+" ");

            for(int i=1; i<n+1; i++) {
                if (graph[x][i] == 1 && !visited[i]) {
                    queue.add(i);
                    visited[i] = true;
                }
            }
        }
    }
}